tests/testthat/test-derive_param_response.R

```diff
@@ -1,92 +1,144 @@
-    "3", "2021-10-30", "PD", "Y"
-  ) %>%
-    mutate(
-      STUDYID = "XX1234",
-      ADT = ymd(ADTC),
-    ) %>%
-    select(-ADTC)
+library(tibble)
+library(lubridate)
+library(dplyr)
+library(rlang)
+library(admiraldev)
+library(admiral)
 
+#+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+# Create Test Data ----
+#+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 
-  rsp <- adrs1 %>%
-    filter(PARAMCD == "OVR" & AVALC %in% c("CR", "PR")) %>%
-    left_join(adrs1 %>%
-      filter(PARAMCD == "PD") %>%
-      select(STUDYID, USUBJID, PDDT = ADT),
-    by = c("STUDYID", "USUBJID")
-    ) %>%
-    filter(ADT <= PDDT | is.na(PDDT)) %>%
-    arrange(USUBJID, PARAMCD, ADT) %>%
-    group_by(USUBJID, PARAMCD) %>%
-    filter(row_number() == 1) %>%
-    mutate(avalc = 1)
+adsl <- tribble(
+  ~USUBJID, ~TRTSDTC,      ~CHECKKEPTCOL, ~CHECKNOTKEPTCOL,
+  "1",      "2020-01-01",  "001",         "991",
+  "2",      "2019-12-12",  "002",         "992",
+  "3",      "2019-11-11",  "003",         "993",
+  "4",      "2019-12-30",  "004",         "994",
+) %>%
+  mutate(
+    TRTSDT = ymd(TRTSDTC),
+    STUDYID = "XX1234"
+  )
+
+adrs <- tribble(
+  ~USUBJID, ~ADTC, ~PARAMCD, ~AVALC,
+  "1", "2020-01-02", "OVR", "PR",
+  "1", "2020-02-01", "OVR", "CR",
+  "1", "2020-03-01", "OVR", "CR",
+  "1", "2020-04-01", "OVR", "SD",
+  "1", NA_character_, "PD", "N",
+  "2", "2021-06-15", "OVR", "SD",
+  "2", "2021-07-16", "OVR", "PD",
+  "2", "2021-09-14", "OVR", "PD",
+  "2", "2021-09-14", "PD", "Y",
+  "3", "2021-09-14", "OVR", "SD",
+  "3", "2021-10-30", "OVR", "PD",
+  "3", "2021-12-25", "OVR", "CR",
+  "3", "2021-10-30", "PD", "Y"
+) %>%
+  mutate(
+    STUDYID = "XX1234",
+    ADT = ymd(ADTC),
+  ) %>%
+  select(-ADTC)
 
-  new_obs <- adsl %>%
-    left_join(rsp, by = c("STUDYID", "USUBJID")) %>%
-    mutate(
-      AVALC = if_else(avalc == 1, "Y", "N", "N"),
-      AVAL = if_else(avalc == 1, 1, 0, 0),
-      PARAMCD = "RSP", PARAM = "Response by investigator"
+# derive_param_response, Test 1 ----
+test_that("Test 1: Test that response is derived accurately, with source_pd", {
+  expected_output <- bind_rows(
+    adrs,
+    tribble(
+      ~USUBJID, ~ADTC,        ~AVALC, ~AVAL, ~TRTSDTC,     ~CHECKKEPTCOL,
+      "1",      "2020-01-02", "Y",    1,     "2020-01-01", "001",
+      "2",      "",           "N",    0,     "2019-12-12", "002",
+      "3",      "",           "N",    0,     "2019-11-11", "003",
+      "4",      "",           "N",    0,     "2019-12-30", "004",
     ) %>%
-    select(-avalc, -PDDT)
+      mutate(
+        ADT = ymd(ADTC),
+        TRTSDT = ymd(TRTSDTC),
+        STUDYID = "XX1234",
+        PARAMCD = "RSP",
+        PARAM = "Response by investigator"
+      ) %>%
+      select(-ADTC, -TRTSDTC)
+  )
 
-  expected_output <- bind_rows(adrs1, new_obs)
+  # Dropping TRTSDT and CHECKKEPTCOL, but I believe the should be in the output dataframe
+  # TBC: sgorm123
```

Comment:
Our convention is to keep the variables from `dataset` if the new observation originates from `dataset`. If the new observation originates from `dataset_adsl`, only those variables which are in both `dataset` and `dataset_adsl` are kept. (See also https://pharmaverse.github.io/admiral/reference/derive_param_first_event.html#details.)

Thus `TRTSDT` and `CHECKKEPTCOL` should not be kept because they are not in `adrs`.

Reply:
ah yes, you are right, I had not put CHECKKEPTCOL nor TRTSTDT onto ADRS - I have done this now and it works.

Pushing change in a minute, thank you.

Reply:
Great, thanks